libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -4465,6 +4468,55 @@ void GCS_MAVLINK::send_set_position_target_global_int(uint8_t target_system, uin
             0,0);   // yaw, yaw_rate
 }
 
+void GCS_MAVLINK::send_smart_battery_info(const uint8_t instance) const
+{
+    const AP_BattMonitor &battery = AP::battery();
+
+	uint16_t cycles;
+	bool got_cycle_count = battery.get_cycle_count(instance,cycles);
+
+	int32_t capacity_full = battery.pack_capacity_mah(instance);
+
+    if (capacity_full == 0) {
+        capacity_full = -1;
+    }
+
+    char serial_number[16]{};
+    battery.get_serial_number(instance, serial_number, ARRAY_SIZE(serial_number)); 
+
+    char product_name[50]{}; 
+    battery.get_product_name(instance, product_name, ARRAY_SIZE(product_name));
+
+    mavlink_msg_smart_battery_info_send(
+        chan,
+        instance,                       // Battery ID
+        MAV_BATTERY_FUNCTION_UNKNOWN,   // (not provided) Function of the battery
+        MAV_BATTERY_TYPE_UNKNOWN,       // (not provided) Type (chemistry) of the battery
+        -1,                             // (not provided) [mAh] Capacity when full according to manufacturer, -1: field not provided.
+        capacity_full,                  // [mAh] Capacity when full (accounting for battery degradation), -1: field not provided.
+        got_cycle_count ? cycles : UINT16_MAX,  // Charge/discharge cycle count. UINT16_MAX: field not provided.
+        serial_number,                  // Serial number in ASCII characters, 0 terminated. All 0: field not provided.
+        product_name,                   // Static device name. Encode as manufacturer and product names separated using an underscore.
+        0,                              // (not provided) [g] Battery weight. 0: field not provided.                                  
+        UINT16_MAX,                     // (not provided) [mV] Minimum per-cell voltage when discharging. If not supplied set to UINT16_MAX value.  
+        UINT16_MAX,                     // (not provided) [mV] Minimum per-cell voltage when charging. If not supplied set to UINT16_MAX value.
+        UINT16_MAX);                    // (not provided) [mV] Minimum per-cell voltage when resting. If not supplied set to UINT16_MAX value.
+}
+
+// returns true if all battery instances were reported
+bool GCS_MAVLINK::send_smart_battery_info() const
+{
+    const AP_BattMonitor &battery = AP::battery();
+    
+    for(uint8_t i = 0; i < battery.num_instances(); i++) {
+        if (battery.get_type(i) != AP_BattMonitor::Type::NONE) {
+            CHECK_PAYLOAD_SIZE(SMART_BATTERY_INFO);
+            send_smart_battery_info(i);
+        }
```

Comment:
@WickedShell has grown to dislike this pattern, preferring to keep some state around indicating the last instance we successfully sent.

Reply:
The pattern for the other battery message like you describe has to be modified with the comment below due to where I have the message being sent from. Suggestions?

    for(uint8_t i = 0; i < battery.num_instances(); i++) {
        const uint8_t battery_id = (last_smart_battery_info_idx + 1) % AP_BATT_MONITOR_MAX_INSTANCES;
        if (battery.get_type(battery_id) != AP_BattMonitor::Type::NONE) {
            CHECK_PAYLOAD_SIZE(SMART_BATTERY_INFO);
            send_smart_battery_info(battery_id);
            last_smart_battery_info_idx = battery_id;
            //return true;
        } else {
            last_smart_battery_info_idx = battery_id;
        }
    }
    return true;